ably/realtime_conn.go

```diff
@@ -385,21 +385,41 @@ func (c *Connection) connectWith(arg connArgs) (result, error) {
 			ConnectionStateDisconnected,
 		)
 	}
-	connectMode := c.getMode()
-	query, err := c.params(connectMode)
-	if err != nil {
-		return nil, err
-	}
-	u.RawQuery = query.Encode()
-	proto := c.opts.protocol()
 
-	if c.State() == ConnectionStateClosed { // RTN12d - if connection is closed by client, don't try to reconnect
-		return nopResult, nil
-	}
+	var conn conn
+	primaryHost := c.opts.getRealtimeHost()
+	fallbackHosts, _ := c.opts.getFallbackHosts()
+	// Always try primary host first and then fallback hosts for realtime conn
+	hosts := append([]string{primaryHost}, ablyutil.Shuffle(fallbackHosts)...)
+	for hostCounter, host := range hosts {
+		u, err := url.Parse(c.opts.realtimeURL(host))
+		if err != nil {
+			return nil, err
+		}
+		query, err := c.params(connectMode)
+		if err != nil {
+			return nil, err
+		}
+		u.RawQuery = query.Encode()
+		proto := c.opts.protocol()
 
-	// if err is nil, raw connection with server is successful
-	conn, err := c.dial(proto, u)
-	if err != nil {
+		if c.State() == ConnectionStateClosed { // RTN12d - if connection is closed by client, don't try to reconnect
+			return nopResult, nil
+		}
+		// if err is nil, raw connection with server is successful
+		conn, err = c.dial(proto, u)
+		if err == nil { // success
```

Comment:
go programmers are more used to reading the error being handled inside the if statement, and the non-error case afterwards.

Reply:
updated